internal/dataplane/parser/parser.go

```diff
@@ -386,14 +390,134 @@ func getCertFromSecret(secret *corev1.Secret) (string, string, error) {
 	return cert, key, nil
 }
 
-func getCerts(log logrus.FieldLogger, s store.Storer, secretsToSNIs map[string][]string) []kongstate.Certificate {
-	snisAdded := make(map[string]bool)
-	// map of cert public key + private key to certificate
-	type certWrapper struct {
-		cert              kong.Certificate
-		CreationTimestamp metav1.Time
+type certWrapper struct {
+	identifier        string
+	cert              kong.Certificate
+	snis              []string
+	CreationTimestamp metav1.Time
+}
+
+func getGatewayCerts(log logrus.FieldLogger, s store.Storer) []certWrapper {
+	certs := []certWrapper{}
+	gateways, err := s.ListGateways()
+	if err != nil {
+		log.WithError(err).Error("failed to list Gateways")
+		return certs
+	}
+	policies, err := s.ListReferencePolicies()
+	if err != nil {
+		log.WithError(err).Error("failed to list ReferencePolicies")
+		return certs
 	}
-	certs := make(map[string]certWrapper)
+	for _, g := range gateways {
+		statuses := make(map[gatewayv1alpha2.SectionName]gatewayv1alpha2.ListenerStatus, len(g.Status.Listeners))
+		for _, status := range g.Status.Listeners {
+			statuses[status.Name] = status
+		}
+		for _, l := range g.Spec.Listeners {
+			ready := false
+			if status, ok := statuses[l.Name]; ok {
+				log.WithFields(logrus.Fields{
+					"gateway":  g.Name,
+					"listener": l.Name,
+				}).Debug("listener missing status information")
+				for _, c := range status.Conditions {
+					if c.Type == string(gatewayv1alpha2.ListenerConditionReady) {
+						if c.Status == metav1.ConditionTrue {
+							ready = true
+						}
+					}
+				}
+			}
+			if ready {
```

Comment:
Renamed things. For the loop, I think we should avoid various single-purpose single-type functions and instead generically handle finding a Condition with certain qualities, since we do that everywhere. Are you okay deferring that to https://github.com/Kong/kubernetes-ingress-controller/issues/2609 ?

Reply:
> Renamed things. For the loop, I think we should avoid various single-purpose single-type functions and instead generically handle finding a Condition with certain qualities

Can you help me to understand better why you feel we should avoid this? I don't fundamentally think there's any problem with single-purpose/single-type functions and in fact I think they can significantly improve composibility, readability, and testability of code.

> since we do that everywhere.

I guess what I'm advocating for then is to change the style here where we have a chance, rather than continuing to commit to an existing style that I think isn't as optimal.

Reply:
I take it that your concern is if we do it here, then when we tried to do the bigger refactor we would somehow miss it. I feel uncertain about that stance, but at the same time this is not something that I feel needs to block this PR so in the spirit of keeping things moving and because we have a follow up issue now to look at it more holistically, let's defer to #2609 :+1: